src/tiledb/cloud/geospatial/ingestion.py

```diff
@@ -0,0 +1,1287 @@
+    with tiledb.scope_ctx(config):
+        with Profiler(array_uri=log_uri, id=id, trace=trace):
+            vfs = tiledb.VFS()
+            logger = get_logger_wrapper(verbose)
+
+            def scale_array(arr):
+                for idx, d in enumerate(["X", "Y", "Z"]):
+                    arr[d] = offsets[idx] + (arr[d] * scales[idx])
+                return arr
+
+            if append:
+                for f in sources:
+                    logger.debug("ingesting point cloud %r", f)
+                    with vfs.open(f, "rb") as src:
+                        las = laspy.open(src)
+                        chunk_itr = las.chunk_iterator(chunk_size)
+
+                        for chunk in chunk_itr:
+                            arr = scale_array(chunk.array)
+                            pipeline = pdal.Writer.tiledb(
+                                array_name=dataset_uri, append=append
+                            ).pipeline(arr)
+                            pipeline.execute()
+            else:
+                if extents and offsets and scales and template_sample:
+                    logger.debug("creating point cloud schema from %r", template_sample)
+                    with vfs.open(template_sample, "rb") as src:
+                        las = laspy.open(src)
+                        chunk_itr = las.chunk_iterator(chunk_size)
+                        first_chunk = next(chunk_itr)
+                        arr = scale_array(first_chunk.array)
+
+                        pipeline = pdal.Writer.tiledb(
+                            array_name=dataset_uri,
+                            x_domain_st=extents.minx - 1,
+                            y_domain_st=extents.miny - 1,
+                            z_domain_st=extents.minz - 1,
+                            x_domain_end=extents.maxx + 1,
+                            y_domain_end=extents.maxy + 1,
+                            z_domain_end=extents.maxz + 1,
+                            offset_x=offsets[0],
+                            offset_y=offsets[1],
+                            offset_z=offsets[1],
+                            scale_x=scales[0],
+                            scale_y=scales[1],
+                            scale_z=scales[1],
+                            chunk_size=chunk_size,
+                        ).pipeline(arr)
+                        pipeline.execute()
+
+                        for chunk in chunk_itr:
+                            arr = scale_array(chunk.array)
+                            pipeline = pdal.Writer.tiledb(
+                                array_name=dataset_uri, append=True
+                            ).pipeline(arr)
+                            pipeline.execute()
+                else:
+                    raise ValueError("Insufficient metadata for point cloud ingestion")
+
+            logger.info("max memory usage: %.3f GiB", max_memory_usage() / (1 << 30))
+
+
+def ingest_raster_udf(
+    chunks: Optional[Tuple[GeoBlockMetadata, ...]] = None,
```

Comment:
In general, for inputs to functions, it's usually better to take in a general Iterable rather than a variable-sized tuple. If you need it to be a tuple itself, you can convert it (which, if it's already a tuple, is free because it will give you back the same object).

Reply:
👍